presto-main/src/main/java/io/prestosql/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -640,7 +647,7 @@ public PlanWithProperties visitProject(ProjectNode node, HashComputationSet pare
                     hashExpression = hashSymbol.toSymbolReference();
                 }
                 newAssignments.put(hashSymbol, hashExpression);
-                allHashSymbols.put(hashComputation, hashSymbol);
```

Comment:
So the issue was that we weren't translating from output->input and recording the hash computation with references to the output symbols? I'm surprised it's not failing in other ways.

Reply:
It happens only for `Union All` queries where the input and output were just a mapping from its source (not derived from same symbol) and it results in recomputing the hash once again